src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs

```diff
@@ -0,0 +1,1237 @@
+            var verifier = CompileAndVerify(comp, expectedOutput: "42");
+            verifier.VerifyIL("S.M", @"
+{
+  // Code size        7 (0x7)
+  .maxstack  1
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""int S.Property.get""
+  IL_0006:  ret
+}
+");
+        }
+
+        [Fact]
+        public void TypeDeclaration_NoBaseInitializer()
+        {
+            var src = @"
+public record struct S
+{
+    public S(int i) : base() { }
+}
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (4,12): error CS0522: 'S': structs cannot call base class constructors
+                //     public S(int i) : base() { }
+                Diagnostic(ErrorCode.ERR_StructWithBaseConstructorCall, "S").WithArguments("S").WithLocation(4, 12)
+                );
+        }
+
+        [Fact]
+        public void TypeDeclaration_NoParameterlessConstructor()
+        {
+            var src = @"
+public record struct S
+{
+    public S() { }
+}
+";
+            // PROTOTYPE(record-structs): this will be allowed in C# 10
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (4,12): error CS0568: Structs cannot contain explicit parameterless constructors
+                //     public S() { }
+                Diagnostic(ErrorCode.ERR_StructsCantContainDefaultConstructor, "S").WithLocation(4, 12)
+                );
+        }
+
+        [Fact]
+        public void TypeDeclaration_NoInstanceInitializers()
+        {
+            var src = @"
+public record struct S
+{
+    public int field = 42;
+    public int Property { get; set; } = 43;
+}
+";
+            // PROTOTYPE(record-structs): this will be allowed in C# 10
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (4,16): error CS0573: 'S': cannot have instance property or field initializers in structs
+                //     public int field = 42;
+                Diagnostic(ErrorCode.ERR_FieldInitializerInStruct, "field").WithArguments("S").WithLocation(4, 16),
+                // (5,16): error CS0573: 'S': cannot have instance property or field initializers in structs
+                //     public int Property { get; set; } = 43;
+                Diagnostic(ErrorCode.ERR_FieldInitializerInStruct, "Property").WithArguments("S").WithLocation(5, 16)
+                );
+        }
+
+        [Fact]
+        public void TypeDeclaration_NoDestructor()
+        {
+            var src = @"
+public record struct S
+{
+    ~S() { }
+}
+";
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
+                // (4,6): error CS0575: Only class types can contain destructors
+                //     ~S() { }
+                Diagnostic(ErrorCode.ERR_OnlyClassesCanContainDestructors, "S").WithArguments("S.~S()").WithLocation(4, 6)
+                );
+        }
+
+        [Fact]
+        public void TypeDeclaration_DifferentPartials()
```

Comment:
>TypeDeclaration_DifferentPartials [](start = 20, length = 33)

It feels like there are more permutations that we can test. Perhaps we should cover `record struct` declaration paired with all other kinds of declarations, the same for `record class`.